File: 2024/day05/main.py
from functools import cmp_to_key

chunks = open('input').read().split('\n\n')
chunks = [chunk.splitlines() for chunk in chunks]

rules = {}
for line in chunks[0]:
	l, r = line.split('|')
	if l not in rules:
		rules[l] = []
	rules[l].append(r)

updates = [line.split(',') for line in chunks[1]]


def cmp(a, b):
	return -1 if b in rules[a] else 1


def check(update):
	for i in range(1, len(update)):
		if cmp(update[i - 1], update[i]) > 0:
			return False
	return True


# Part 1: 4609
sum = 0
for u in updates:
	if check(u):
		mid = len(u) // 2
		sum += int(u[mid])
print(sum)

# Part 2: 5723
sum = 0
for u in updates:
	if not check(u):
		u.sort(key=cmp_to_key(cmp))
		mid = len(u) // 2
		sum += int(u[mid])
print(sum)

File: 2024/day05/main.js
const { readFileSync } = require('fs');

const chunks = readFileSync('input', 'utf-8')
  .trim()
  .split('\n\n')
  .map((chunk) => chunk.split('\n'));

const rules = {};
chunks[0].forEach((line) => {
  const [l, r] = line.split('|');
  if (!rules[l]) {
    rules[l] = [];
  }
  rules[l].push(r);
});

const updates = chunks[1].map((line) => line.split(','));

function cmp(a, b) {
  return rules[a].includes(b) ? -1 : 1;
}

function check(u) {
  for (let i = 1; i < u.length; i++) {
    if (cmp(u[i - 1], u[i]) > 0) {
      return false;
    }
  }
  return true;
}

// Part 1: 4609
let sum = 0;
updates.forEach((u) => {
  if (check(u)) {
    const mid = Math.floor(u.length / 2);
    sum += parseInt(u[mid], 10);
  }
});
console.log(sum);

// Part 1: 5723
sum = 0;
updates.forEach((u) => {
  if (!check(u)) {
    u.sort(cmp);
    const mid = Math.floor(u.length / 2);
    sum += parseInt(u[mid], 10);
  }
});
console.log(sum);
